# Snakemake Partial Genome Sequencing Pipeline

Pipeline for processing Illumina sequencing data generated by target enrichment via hybrid capture experiments. 
Heavily follows the Phyluce methodology outlined in 
[Tutorial I: UCE Phylogenomics](https://phyluce.readthedocs.io/en/latest/tutorial-one.html). 

1) Trims Illumina adapters and bases below <10 quality score [BBDuk](https://jgi.doe.gov/data-and-tools/bbtools/bb-tools-user-guide/bbduk-guide/)
2) Assembles trimmed reads [Abyss](http://www.bcgsc.ca/platform/bioinfo/software/abyss), [SPAdes, rnaSPAdes](http://cab.spbu.ru/software/spades/)
3) Detects and extracts target contigs [Phyluce](https://phyluce.readthedocs.io/en/latest/index.html) 
4) Summary statistics on targets and assemblies [BBTools Stats](https://jgi.doe.gov/data-and-tools/bbtools/bb-tools-user-guide/statistics-guide/)
5) Optional Script to proceed with Phylogeny

### Prerequisites

* [Conda](https://conda.io/docs/user-guide/install/index.html)
```
wget https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh
bash Miniconda3-latest-Linux-x86_64.sh
```
* [Snakemake](https://snakemake.readthedocs.io/en/stable/getting_started/installation.html)
```
conda install -c bioconda -c conda-forge snakemake
```
* Git

## Getting Started

Within a working directory:

```
git clone https://github.com/AAFC-BICoE/snakemake-partial-genome-pipeline.git .
```
* Create a folder named "fastq" that contains Illumina based raw reads in fastq.gz format. Fastq files should not begin
with numbers, or contain a mix of "_" and "-" characters. 
* Create a folder named "probes" that contains a probe fasta file with fasta headers in 
[Phyluce UCE format](https://phyluce.readthedocs.io/en/latest/uce-processing.html#match-contigs-to-probes)
```
>uce-1_p1
GCTGGTTATC...
>uce-1_p2
TAACAATA....
>uce-2_p1
AAGCATCT...
```
Pipeline split into two snakefiles. "Snakefile" handles the SPAdes and rnaSPAdes pathways. However, in testing it was
found that automating abyss on AAFC cluster computers was proving challenging, so the Abyss functionality was moved
into a seperate snakefile. After the first snakefile successfully completes, run the second snakefile to complete
Abyss and merge all results together. 

Dry-run to see if everything prepared correctly
```
snakemake --use-conda -n
```
To run pipeline with 32 cores and continue if some samples fail:
```
snakemake --use-conda -k --cores 32 
```
Run Abyss pipeline
```bash
snakemake --use-conda -k --cores 32 -s abyss_snakefile
```

## Copyright
Government of Canada, Agriculture & Agri-Food Canada

## License
This project is licensed under the MIT License - see the [LICENSE](LICENSE) file for details

## Citations

* [BioPython](https://biopython.org/) - Tools for biological computation  
Cock, P.J.A. et al. Biopython: freely available Python tools for computational molecular biology and bioinformatics. 
Bioinformatics 2009 Jun 1; 25(11) 1422-3 http://dx.doi.org/10.1093/bioinformatics/btp163 pmid:19304878

* [Snakemake](https://snakemake.readthedocs.io/en/stable/) - Workflow management system    
Köster, Johannes and Rahmann, Sven. “Snakemake - A scalable bioinformatics workflow engine”. Bioinformatics 2012.

* SPAdes  
Nurk S. et al. (2013) Assembling Genomes and Mini-metagenomes from Highly Chimeric Reads. In: Deng M., Jiang R., 
Sun F., Zhang X. (eds) Research in Computational Molecular Biology. RECOMB 2013. Lecture Notes in Computer Science, 
vol 7821. Springer, Berlin, Heidelberg

* BBTools  
Brian-JGI (2018) BBTools is a suite of fast, multithreaded bioinformatics tools designed for analysis of DNA and RNA 
sequence data.https://jgi.doe.gov/data-and-tools/bbtools/ 

* FASTQC  
Andrews S. (2018). FastQC: a quality control tool for high throughput sequence data. 
Available online at: http://www.bioinformatics.babraham.ac.uk/projects/fastqc

* [Phyluce](https://phyluce.readthedocs.io/en/latest/index.html) - Target enrichment data analysis    
Faircloth BC. 2016. PHYLUCE is a software package for the analysis of conserved genomic loci. 
Bioinformatics 32:786-788. doi:10.1093/bioinformatics/btv646.

* Ultraconserved elements  
BC Faircloth, McCormack JE, Crawford NG, Harvey MG, Brumfield RT, Glenn TC. 2012. Ultraconserved elements anchor 
thousands of genetic markers spanning multiple evolutionary timescales. Systematic Biology 61: 717–726. 
doi:10.1093/sysbio/SYS004.

## Author
Jackson Eyres \
Bioinformatics Programmer \
Agriculture & Agri-Food Canada \
jackson.eyres@canada.ca

