# Snakemake Partial Genome Sequencing Pipeline

Pipeline for assembling and analysing sequencing data generated by target enrichment via hybrid capture. Trims and assembles raw reads and processes the detection of targets using the Phyluce workflow (Reference) 

## Getting Started

Within a working directory, pipeline expects two directories:
1) fastq: Contains Illumina based raw reads in fastq 
2) probes/<probe file in Phyluce UCE format>
```
cd <directory>
snakemake --use-conda
```
These instructions will get you a copy of the project up and running on your local machine for development and testing purposes. See deployment for notes on how to deploy the project on a live system.

### Prerequisites

* [Conda Installation](https://conda.io/docs/user-guide/install/index.html)
* [Snakemake Installation](https://snakemake.readthedocs.io/en/stable/getting_started/installation.html)

```
wget https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh
bash Miniconda3-latest-Linux-x86_64.sh
conda install -c bioconda -c conda-forge snakemake
```

### Installing

Install pipeline files into the working directory of the data

```
git clone https://github.com/AAFC-BICoE/snakemake-partial-genome-pipeline.git
```

### Examples

Add additional notes about how to deploy this on a live system

## Built With

* [Python](https://www.python.org/doc/) - Programming language
* [Conda](https://conda.io/docs/index.html) - Package, dependency and environment management
* [Snakemake](https://snakemake.readthedocs.io/en/stable/) - Workflow management system
* [Phyluce](https://phyluce.readthedocs.io/en/latest/index.html) - Target enrichment data analysis
* [BioPython](https://biopython.org/) - Tools for biological computation

## Authors

* **Jackson Eyres** email: jackson.eyres@canada.ca

## License

This project is licensed under the MIT License - see the [LICENSE](LICENSE) file for details

